rllib/env/remote_base_env.py

```diff
@@ -0,0 +1,225 @@
+                repeat the ray.wait call until at least one sub-env is
+                ready.
+            existing_envs: Optional list of already created sub-environments.
+                These will be used as-is and only as many new sub-envs as
+                necessary (`num_envs - len(existing_envs)`) will be created.
+        """
+
+        # Could be creating local or remote envs.
+        self.make_env = make_env
+        # Whether the given `make_env` callable already returns ray.remote
+        # objects or not.
+        self.make_env_creates_actors = False
+        # Already existing env objects (generated by the RolloutWorker).
+        self.existing_envs = existing_envs or []
+        self.num_envs = num_envs
+        self.multiagent = multiagent
+        self.poll_timeout = remote_env_batch_wait_ms / 1000
+
+        self.actors = None  # lazy init
+        self.pending = None  # lazy init
+
+    @override(BaseEnv)
+    def poll(self) -> Tuple[MultiEnvDict, MultiEnvDict, MultiEnvDict,
+                            MultiEnvDict, MultiEnvDict]:
+        if self.actors is None:
+            # `self.make_env` already produces Actors: Use it directly.
+            if len(self.existing_envs) > 0 and isinstance(
+                    self.existing_envs[0], ray.actor.ActorHandle):
+                self.make_env_creates_actors = True
+                self.actors = []
+                while len(self.actors) < self.num_envs:
+                    self.actors.append(self.make_env(len(self.actors)))
+            # `self.make_env` produces gym.Envs (or children thereof, such
+            # as MultiAgentEnv): Need to auto-wrap it here. The problem with
+            # this is that custom methods wil get lost. If you would like to
+            # keep your custom methods in your envs, you should provide the
+            # env class directly in your config (w/o tune.register_env()),
+            # such that your class will directly be made a @ray.remote
+            # (w/o the wrapping via `_Remote[Multi|Single]AgentEnv`).
+            else:
+
+                def make_remote_env(i):
+                    logger.info("Launching env {} in remote actor".format(i))
+                    if self.multiagent:
+                        return _RemoteMultiAgentEnv.remote(self.make_env, i)
+                    else:
+                        return _RemoteSingleAgentEnv.remote(self.make_env, i)
+
+                self.actors = [
+                    make_remote_env(i) for i in range(self.num_envs)
+                ]
+
+        if self.pending is None:
+            self.pending = {a.reset.remote(): a for a in self.actors}
+
+        # each keyed by env_id in [0, num_remote_envs)
+        obs, rewards, dones, infos = {}, {}, {}, {}
+        ready = []
+
+        # Wait for at least 1 env to be ready here
+        while not ready:
+            ready, _ = ray.wait(
+                list(self.pending),
+                num_returns=len(self.pending),
```

Comment:
I don't think that's true. In my understanding, ray.wait returns a tuple: 1) list of ready-handles and 2) list of non-ready handles. As soon as there is at least one handle ready, we continue and return observations from only those ready ray.remote sub-environments.

Reply:
ok, yeah, that's the self.pending thing I guess. let me double check the semantics for num_wait.

In any case though, I think we should get rid of the while loop and the time_out parameter, and this is the same thing right?

Reply:
Ah, yeah, you are right! No wait + no timeout == what we are currently doing (but much simpler ;)).

Reply:
ok I verified a couple of things and understand this completely now :)
so normally num_returns=len(self.pending) would wait until all pending actors come back, however we also have timeout=self.poll_timeout there ... so effectively the logic here is:

1. we wait self.poll_timeout, if all workers come back with a batch in time, great, and continue.
2. if some workers are slow, and don't come back in poll_timeout, we leave them, and continue, so we don't wait more than poll_timeout interval.
3. if all workers are slow, and nothing comes back in poll_timeout though, we will ignore poll_timeout and keep waiting until at least something comes back, because of the while loop. this may or may not be an intended behavior actually, for off policy cases?

I think as we discussed offline, let's just comment about this behavior and leave it as is for now. this sounds like a performance impacting change to me, and we should do it by itself I guess.
I am also thinking if we are ok with getting data back from some of the workers, then we don't really need to have the users specify this parameter, which will be 1 fewer confusing parameter for the users :)

but, yeah, we can come back to this. thanks :)

Reply:
1. 2. and 3.: Yeah, that's how I understood it now, too. So removing the timeout would actually alter the logic here.